File: src/com/cowlark/stellation2/server/model/SPlayer.java
/* Server-side player.
 * $Source: /cvsroot/stellation/stellation2/src/com/cowlark/stellation2/server/model/SPlayer.java,v $
 * $Date: 2009/09/15 23:15:49 $
 * $Author: dtrg $
 * $Revision: 1.3 $
 */

package com.cowlark.stellation2.server.model;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.TreeMap;
import com.cowlark.stellation2.common.Resources;
import com.cowlark.stellation2.common.S;
import com.cowlark.stellation2.common.UpdateBatch;
import com.cowlark.stellation2.common.db.ListOfClientObjects;
import com.cowlark.stellation2.common.db.LogMessageStore;
import com.cowlark.stellation2.common.exceptions.FleetAlreadyHasThisNameException;
import com.cowlark.stellation2.common.exceptions.NotAdministratorException;
import com.cowlark.stellation2.common.exceptions.StellationException;
import com.cowlark.stellation2.common.model.CObject;
import com.cowlark.stellation2.common.model.CPlayer;
import com.cowlark.stellation2.server.Log;
import com.cowlark.stellation2.server.db.CClass;
import com.cowlark.stellation2.server.db.DataMangler;
import com.cowlark.stellation2.server.db.MapOfServerObjects;
import com.cowlark.stellation2.server.db.Property;
import com.cowlark.stellation2.server.db.ServerDB;

@CClass(name = CPlayer.class)
public class SPlayer extends SObject
{
    private static final long serialVersionUID = -954295233192947332L;

	@Property(scope = S.GLOBAL)
    private String _uid;

    @Property
    private String _password;
    
    @Property(scope = S.GLOBAL)
    private String _name;

    @Property(scope = S.GLOBAL)
    private String _empire;

    @Property(scope = S.OWNER)
    private boolean _administrator;
    
    @Property(scope = S.OWNER)
    private MapOfServerObjects<SFleet> _fleets = new MapOfServerObjects<SFleet>();
    
    @Property
    private LogMessageStore _log = new LogMessageStore();
    
    private Map<Long, Integer> _visibleObjects = new TreeMap<Long, Integer>();

    public SPlayer initialise(String uid, String password, String empire, String name)
    {
    	super.initialise();
    	
    	_uid = uid;
    	_password = password;
    	_name = name;
    	_empire = empire;
    	_administrator = true;
    	
    	setOwner(this);
    	
    	SStar star = getGalaxy().getRandomStar();
    	
    	try
    	{
	    	SFleet fleet = createFleet(star, _empire+"'s starter fleet");
	    	
	    	fleet.createJumpship()
	    		.makeAlive();
	    	
	    	fleet.createTug()
	    		.makeAlive()
	    		.createBasicFactory();
	    	
	    	fleet.createCargoship()
	    		.setCargo(new Resources(100000.0, 100000.0, 100000.0))
	    		.makeAlive();
	    	
	    	log("Intergalactic transit bubble arrival. " + _empire +
	    			" of " + _name + " has arrived in the galaxy.");
    	}
    	catch (StellationException e)
    	{
    	}
    	
    	return this;
    }
    
	public String getUid()
    {
    	return _uid;
    }

	public SPlayer setUid(String uid)
    {
    	_uid = uid;
    	dirty();
    	return this;
    }

	public String getPassword()
	{
		return _password;
	}
	
	public void setPassword(String password)
	{
		_password = password;
	}
	
    public String getName()
    {
    	return _name;
    }

	public SPlayer setName(String name)
    {
    	_name = name;
    	dirty();
    	return this;
    }

	public String getEmpire()
    {
    	return _empire;
    }

	public SPlayer setEmpire(String empire)
    {
    	_empire = empire;
    	dirty();
    	return this;
    }

	public MapOfServerObjects<SFleet> getFleets()
    {
    	return _fleets;
    }

	public SPlayer setFleets(MapOfServerObjects<SFleet> fleets)
    {
    	_fleets = fleets;
    	dirty();
    	return this;
    }

    public Map<Long, Integer> getVisibleObjects()
    {
	    return _visibleObjects;
    }
    
    public void checkAdministrator() throws NotAdministratorException
    {
    	if (!_administrator)
    		throw new NotAdministratorException();
    }
    
	public SFleet createFleet(SStar star, String name)
			throws FleetAlreadyHasThisNameException
    {
    	if (_fleets.containsKey(name))
    		throw new FleetAlreadyHasThisNameException(name);
    	
    	SFleet fleet = new SFleet();
    	fleet.initialise(star, this, name);
    	_fleets.put(name, fleet);
    	dirty();
    	
    	return fleet;
    }
    
    public void renameFleet(SFleet fleet, String name)
    		throws FleetAlreadyHasThisNameException
    {
    	if (fleet.getName().equals(name))
    		return;
    	
    	if (_fleets.containsKey(name))
    		throw new FleetAlreadyHasThisNameException(name);
    	
    	_fleets.remove(fleet.getName());
    	_fleets.put(name, fleet);
    	
    	dirty();
    }
    
    private void addWithContents(SObject object)
    {
		_visibleObjects.put(object.getId(), S.OWNER);
		for (SObject o : object)
			addWithContents(o);
    }
    
    public void updateVisibility()
    {
    	if (_visibleObjects == null)
    	    _visibleObjects = new HashMap<Long, Integer>();
    	else
    		_visibleObjects.clear();
    	
    	/* The player and the universe itself are visible! */
    	
    	_visibleObjects.put(getId(), S.OWNER);
    	_visibleObjects.put(S.UNIVERSE, S.GLOBAL);
    	
    	/* The galaxy and all (glowing) stars are visible. */
    	
    	SGalaxy galaxy = getGalaxy();
    	_visibleObjects.put(galaxy.getId(), S.GLOBAL);
    	for (Long id : galaxy.getVisibleStars().getIds())
    	{
   			_visibleObjects.put(id, S.GLOBAL);
    	}

    	/* We can see where each fleet that has a jumpship is. */
    	
    	for (SObject o : _fleets.values())
    	{
    		SFleet fleet = o.toFleet();
    		if (fleet.getJumpshipCount() > 0)
    		{
	    		/* We can see both the fleet and its star. */
	    		
	    		_visibleObjects.put(fleet.getId(), S.OWNER);
	    		_visibleObjects.put(fleet.getLocation().getId(), S.LOCAL);
	    		
	    		/* We can also see the contents of the fleet. */
	    		
	    		for (SObject oo : fleet)
	    			addWithContents(oo);
    		}
    	}
    	
    	dirty();
    }
    
    public void addChangedItemsToBatch(UpdateBatch batch, long since)
    {
    	long now = System.currentTimeMillis();
    	ServerDB db = ServerDB.Instance;
    	for (long id : _visibleObjects.keySet())
    	{
    		SObject object = (SObject) db.getImpl(id);
    		if (object.getChangedTime() > since)
    		{
    			int scope = _visibleObjects.get(id);
    			CObject cobject = DataMangler.export(object, scope);
    			cobject.setId(id);
    			cobject.setScope(scope);
    			//cobject = object.createPublic(scope);
    			
    			switch (scope)
    			{
    				case S.OWNER:
    				case S.LOCAL:
    					cobject.setContents(
    							(ListOfClientObjects) object.getContents().getClient());
    					break;
    					
    				case S.GLOBAL:
    				{
    					ListOfClientObjects cset = new ListOfClientObjects();
    					
    					HashSet<Long> childids = new HashSet<Long>(object.getContents().getIds());
    					childids.retainAll(_visibleObjects.keySet());

    					for (long childid : childids)
    						cset.add(childid);
		    			
		    			cobject.setContents(cset);
    				}
    			}
    	    	
    			batch.add(cobject);
    		}
    	}
    	
    	batch.getVisibleObjects().addAll(_visibleObjects.keySet());
    	long delta = System.currentTimeMillis() - now;
    	Log.log("batched %d objects out of %d in %d ms (%d ms per object)",
    			batch.getUpdatedObjects().size(),
    			_visibleObjects.size(),
    			delta,
    			delta / batch.getUpdatedObjects().size());

    	now = SUniverse.getStaticUniverse().getCurrentTime();
    	batch.setLogUpdates(_log.getSubset(since, now));
    	batch.setTime(now);
    }
    
    public SPlayer log(String message)
    {
    	long t = SUniverse.getStaticUniverse().getCurrentTime();
    	_log.add(t, message);
    	
//    	JID jid = new JID("david.given@gmail.com");
//    	Message msg = new MessageBuilder()
//    		.withRecipientJids(jid)
//    		.withFromJid(new JID("stellation2@appspot.com"))
//    		.withBody(Utils.renderTime(t) + ": " + message)
//    		.build();
//    	
//    	XMPPService xmpp = XMPPServiceFactory.getXMPPService();
//    	xmpp.sendMessage(msg);
    	
    	return this;
    }
}

File: src/server/model/SPlayer.cc
#include "globals.h"
#include "Log.h"
#include "SUniverse.h"
#include "SGalaxy.h"
#include "SStar.h"
#include "SFleet.h"
#include "SPlayer.h"

SPlayer::SPlayer(Database::Type oid):
	SObject(oid),
	SPlayerProperties(oid)
{
}

SFleet* SPlayer::CreateFleet(SStar* location, const string& name)
{
	for (Datum::ObjectSet::const_iterator i = Fleets->SetBegin(),
			e = Fleets->SetEnd(); i != e; i++)
	{
		SFleet* fleet = SFleet::Get(*i);
		if (*fleet->Name == name)
			throw Hash::FleetAlreadyHasThisName;
	}

	SFleet* fleet = SFleet::Create(*this);

	*fleet->Name = name;
	location->Add(fleet);

	Fleets->AddToSet(fleet);

	return fleet;
}

static void add_with_contents(SPlayer::VisibilityMap& visible, SObject* o)
{
	visible[*o] = Database::LocalVisibility;

	for (SObject::ObjectSet::const_iterator i = o->Contents->SetBegin(),
			e = o->Contents->SetEnd(); i != e; i++)
	{
		SObject* child = SObject::Get(*i);
		add_with_contents(visible, child);
	}
}

void SPlayer::CalculateVisibleObjects(SPlayer::VisibilityMap& visible)
{
	/* The player and the universe itself are visible. */

	SUniverse* universe = SUniverse::Get(Database::Universe);

	visible[*this] = Database::LocalVisibility;
	visible[*universe] = Database::GlobalVisibility;

	/* The galaxy and all (glowing) stars are visible. */

	SGalaxy* galaxy = SGalaxy::Get(universe->Galaxy);
	visible[*galaxy] = Database::GlobalVisibility;

	for (Datum::ObjectSet::const_iterator i = galaxy->VisibleStars->SetBegin(),
			e = galaxy->VisibleStars->SetEnd(); i != e; i++)
	{
		visible[*i] = Database::GlobalVisibility;
	}

	/* We can see every fleet which has a jumpship. */

	ObjectSet locationsWithAJumpship;
	for (Datum::ObjectSet::const_iterator i = Fleets->SetBegin(),
			e = Fleets->SetEnd(); i != e; i++)
	{
		SFleet* fleet = SFleet::Get(*i);
		if (*fleet->JumpshipCount > 0.0)
			locationsWithAJumpship.insert(*fleet->Location);
	}

	for (ObjectSet::const_iterator i = locationsWithAJumpship.begin(),
			e = locationsWithAJumpship.end(); i != e; i++)
	{
		SStar* star = SStar::Get(*i);
		add_with_contents(visible, star);
	}
}

void SPlayer::AccessRW(Database::Type oid)
{
	if (!DatabaseExists(oid, Hash::Location))
		throw Hash::ObjectDoesNotExist;

	/* Is this object the player? */

	if (oid == *this)
		return;

	/* Is this object in a fleet? */

	while (oid)
	{
		SObject* o = SObject::Get(oid);
		switch (*o->Class)
		{
			case Hash::SFleet:
				/* If this fleet belongs to the player *and* it has a
				 * jumpship, the object is accessible. */

				if (*o->Owner == (Database::Type)*this)
				{
					SFleet* fleet = SFleet::Get(o);
					if (*fleet->JumpshipCount > 0)
						return;
				}
				break;

			case Hash::SStar:
			{
				/* If this star contains at least one fleet that belongs to
				 * the player *and* which has a jumpship, the object is
				 * accessible.
				 */

				for (ObjectSet::const_iterator i = o->Contents->SetBegin(),
						e = o->Contents->SetEnd(); i != e; i++)
				{
					if (SObject::GetClass(*i) == Hash::SFleet)
					{
						SFleet* fleet = SFleet::Get(*i);
						if ((*fleet->Owner == *this) && (*fleet->JumpshipCount > 0))
							return;
					}
				}

				break;
			}

			default:
				break;
		}

		oid = *o->Location;
	}

	/* Nope. Note accessible. */

	throw Hash::PrivilegeViolation;
}
